lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -482,14 +516,146 @@ namespace Intl
         error = UErrorCode::U_ZERO_ERROR;
 
         written = uloc_toLanguageTag(defaultLocaleId, bcp47, ULOC_FULLNAME_CAPACITY, true, &error);
-        ICU_ASSERT_RETURN(error, written > 0 && written < cchLangtag, 0);
+        ICU_ASSERT_RETURN(error, written  > 0 && ((size_t) written) < cchLangtag, 0);
 
         return utf8::DecodeUnitsIntoAndNullTerminateNoAdvance(
             langtag,
             reinterpret_cast<LPCUTF8>(bcp47),
             reinterpret_cast<LPCUTF8>(bcp47 + written)
         );
     }
+
+    bool CollatorGetCollation(_In_z_ const char *langtag, _Out_ char *collation, _In_ size_t cchCollation)
+    {
+        Assert(langtag != nullptr && collation != nullptr && cchCollation >= 0 && cchCollation < ULOC_FULLNAME_CAPACITY);
+        collation[0] = 0;
+
+        bool ret = false;
+
+        UErrorCode error = U_ZERO_ERROR;
+        char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t length = 0;
+        uloc_forLanguageTag(langtag, localeID, ULOC_FULLNAME_CAPACITY, &length, &error);
+        ICU_ASSERT_RETURN(error, length > 0, ret);
+
+        UCollator *collator = ucol_open(localeID, &error);
+        ICU_ASSERT_RETURN(error, true, ret);
+
+        const char *collatorLocaleID = ucol_getLocaleByType(collator, ULOC_VALID_LOCALE, &error);
+        ICU_ASSERT_GOTO(error, true, LCloseCollator);
+
+        char collatorCollation[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t collatorCollationLength = uloc_getKeywordValue(collatorLocaleID, "collation", collatorCollation, _countof(collatorCollation), &error);
+        ICU_ASSERT_GOTO(error, collatorCollationLength >= 0, LCloseCollator);
+        if (collatorCollationLength == 0)
+        {
+            // if the collation value doesn't exist, that is fine
+            goto LSuccess;
+        }
+
+        const char *bcpValue = uloc_toUnicodeLocaleType("collation", collatorCollation);
+        size_t cchBcpValue = strlen(bcpValue);
+        if (cchBcpValue == 0 || cchBcpValue >= cchCollation || memcpy_s(collation, cchCollation, bcpValue, cchBcpValue) != 0)
+        {
+            goto LCloseCollator;
+        }
+        else
+        {
+            collation[cchBcpValue] = 0;
+        }
+
+LSuccess:
+        ret = true;
+
+LCloseCollator:
+        ucol_close(collator);
+        return ret;
+    }
+
+    int CollatorCompare(_In_z_ const char16 *langtag, _In_z_ const char16 *left, _In_ charcount_t cchLeft, _In_z_ const char16 *right, _In_ charcount_t cchRight,
+        _In_ CollatorSensitivity sensitivity, _In_ bool ignorePunctuation, _In_ bool numeric, _In_ CollatorCaseFirst caseFirst, _Out_ HRESULT *hr)
+    {
+        ASSERT_ENUM(CollatorSensitivity, sensitivity);
+        ASSERT_ENUM(CollatorCaseFirst, caseFirst);
+        Assert(langtag != nullptr && left != nullptr && cchLeft >= 0 && right != nullptr && cchRight >= 0 && hr != nullptr);
```

Comment:
> cchLeft >= 0 [](start = 55, length = 13)

charcount_t >= 0 is a NOP because charcount_t is unsigned (uint32). Is this actually even required?

Reply:
No, i believe it came from a previous review that suggested asserting all invariants.